prompb/io/prometheus/client/decoder_test.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2025 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io_prometheus_client //nolint:revive
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"io"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/stretchr/testify/require"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+const (
+	testGauge = `name: "go_build_info"
+help: "Build information about the main Go module."
+type: GAUGE
+metric: <
+  label: <
+    name: "checksum"
+    value: ""
+  >
+  label: <
+    name: "path"
+    value: "github.com/prometheus/client_golang"
+  >
+  label: <
+    name: "version"
+    value: "(devel)"
+  >
+  gauge: <
+    value: 1
+  >
+>
+metric: <
+  label: <
+    name: "checksum"
+    value: ""
+  >
+  label: <
+    name: "path"
+    value: "github.com/prometheus/prometheus"
+  >
+  label: <
+    name: "version"
+    value: "v3.0.0"
+  >
+  gauge: <
+    value: 2
+  >
+>
+
+`
+	testCounter = `name: "go_memstats_alloc_bytes_total"
+help: "Total number of bytes allocated, even if freed."
+type: COUNTER
+unit: "bytes"
+metric: <
+  counter: <
+    value: 1.546544e+06
+    exemplar: <
+      label: <
+        name: "dummyID"
+        value: "42"
+      >
+      value: 12
+      timestamp: <
+        seconds: 1625851151
+        nanos: 233181499
+      >
+    >
+  >
+>
+
+`
+)
+
+func TestMetricStreamingDecoder(t *testing.T) {
+	varintBuf := make([]byte, binary.MaxVarintLen32)
```

Comment:
Maybe it would be more future-proof to test this against client_golang i.e. create a registry, add those metrics there, and simulate a protobuf scrape

Reply:
We do that in multiple places across Prometheus codebase e.g. textparse and scrape so I think we should be ok. WDYT?

Reply:
I also plan to someday write a protoc plugin for this so it's autogenerated.